src/main/java/com/azure/cosmos/samples/distributedbulk/JobRepository.java

```diff
@@ -0,0 +1,318 @@
+package com.azure.cosmos.samples.distributedbulk;
+
+import com.azure.cosmos.CosmosAsyncContainer;
+import com.azure.cosmos.CosmosDiagnosticsThresholds;
+import com.azure.cosmos.CosmosException;
+import com.azure.cosmos.models.CosmosItemRequestOptions;
+import com.azure.cosmos.models.CosmosItemResponse;
+import com.azure.cosmos.models.CosmosQueryRequestOptions;
+import com.azure.cosmos.models.PartitionKey;
+import com.azure.cosmos.models.SqlParameter;
+import com.azure.cosmos.models.SqlQuerySpec;
+import com.azure.cosmos.samples.distributedbulk.model.BatchRecord;
+import com.azure.cosmos.samples.distributedbulk.model.IngestionStatus;
+import com.azure.cosmos.samples.distributedbulk.model.InputFileRecord;
+import com.azure.cosmos.samples.distributedbulk.model.JobRecord;
+import com.azure.cosmos.util.CosmosPagedFlux;
+import com.azure.cosmos.util.CosmosPagedIterable;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class JobRepository {
+    private final static Logger logger = LoggerFactory.getLogger(JobRepository.class);
+
+    private static final CosmosAsyncContainer jobContainer =
+        Configs
+            .getCosmosAsyncClient(Main.JobId + "_JobRepository")
+            .getDatabase(Configs.getCosmosDatabaseName())
+            .getContainer(Configs.getCosmosJobContainerName());
+
+    public static void ensureJobDoesNotExistYet(String jobId) {
+        try {
+            jobContainer.readItem(
+                jobId,
+                new PartitionKey(jobId),
+                ObjectNode.class).block();
+
+            throw new IllegalStateException(
+                "A Job metadata document for job ID "
+                    + jobId
+                    + " already exists. Please use unique job ID values.");
+        } catch (CosmosException cosmosException) {
+            if (cosmosException.getStatusCode() == 404
+                && cosmosException.getSubStatusCode() == 1003) {
+
+                throw new IllegalStateException(
+                    "The container '"
+                        + Configs.getCosmosJobContainerName()
+                        + "' does not exist - you have to create this container "
+                        + "manually before running this sample.");
+            } else if (cosmosException.getStatusCode() == 404
+                && cosmosException.getSubStatusCode() == 0) {
+
+                logger.debug(
+                    "The Job metadata document for job ID {} does not exist yet - which is expected.",
+                    jobId);
+            } else {
+                throw cosmosException;
+            }
+        }
+    }
+
+    public static synchronized BatchRecord findBatchProcessingCandidate(int threadId, String jobId) {
```

Comment:
I wonder if this can be made best-effort if the query fails (to determine offset etc.) - I don't see the try-catch handler for it.